### DISCUSSION TITLE
Programatically selecting table cells, rows, columns

### DISCUSSION THREAD

**Discussion by jondavidpass** [ueberdosis/tiptap#2065] [General Questions] [answered]
Hi there,

Just wondering if anyone has attempted this and has some information or examples.

I can see that the Table plugin has a `setCellSelection` command exposed but it's not mentioned at all in the documentation. I'll have a poke at that and see if I can get it working but any help would be appreciated.

**Comment by jondavidpass** [comment, marked as answer]
Okay, so my solution in the end was the create a NodeView for my table cells—this way it is pretty trivial to get their position with the `getPos` method. I then added commands for selecting a row and column by bringing in some extra ProseMirror utilities.

Here is the code for the extension for posterity:

```
import TableCell from '@tiptap/extension-table-cell';
import { CellSelection } from 'prosemirror-tables';
import { VueNodeViewRenderer } from '@tiptap/vue-3';

import TableCellNodeView from '@/path/to/component';

const CustomTableCell = TableCell.extend({
    addNodeView() {
        return VueNodeViewRenderer(TableCellNodeView);
    },
    addCommands() {
        return {
            selectRow: cell => ({ tr: transaction, dispatch }) => {
                if (dispatch) {
                    const position = transaction.doc.resolve(cell);
                    const rowSelection = CellSelection.rowSelection(position);
                    transaction.setSelection(rowSelection);
                }

                return true;
            },
            selectColumn: cell => ({ tr: transaction, dispatch }) => {
                if (dispatch) {
                    const position = transaction.doc.resolve(cell);
                    const colSelection = CellSelection.colSelection(position);
                    transaction.setSelection(colSelection);
                }

                return true;
            },
        };
    }
});

export default CustomTableCell;
```

Then in your NodeView:
```
    methods: {
        selectRow() {
            this.editor.chain().focus().selectRow(this.getPos()).run();
        },
        selectColumn() {
            this.editor.chain().focus().selectColumn(this.getPos()).run();
        }
    }
```

**Reply by hanspagel** [reply]
Thanks for sharing!

**Comment by kevinmelo** [comment]
In my case i wanted to get the styles, normally you need to do is

``` javascript
this.editor.getAttributes('attribute_name')
```
But with `TableCell` you need to select him

``` javascript
this.editor.getAttributes('tableCell') // this will return the attributes object of the TableCell
```

**Comment by gethari** [comment]
Hi @jondavidpass Where does the `this.getPos()` come from ? Is this a `prosemirror-utility` ? And if not can you please share the code for it ?

## Update

Never mind , I found it [HERE](https://tiptap.dev/guide/node-views/vue/#get-pos)

**Reply by gethari** [reply]
@jondavidpass sorry , I tried this today and I am getting this below error from my `NodeViewComponent`

```js
this.getPos is not a function
```

What I'm trying to do is make a table cell as `selected` when I do a right click on it. Here's how my context-wrapper menu looks. I am running this piece of code

```ts
let position: any = {
      anchorCell: this.getPos(),
    };
this.editor.chain().focus().setCellSelection(position).run();
```

![image](https://user-images.githubusercontent.com/20477956/183824220-8c16ee95-00c3-4b8a-b3d0-629d0e1e7909.png)

**Reply by jondavidpass** [reply]
I'm not sure on that one sorry—my implementation didn't end up using `setCellSelection`.

You might find it more helpful to write your own command that converts the `getPos` result into something that the setCellSelection method can actually use.

**Reply by JersyFernandesJF** [reply]
yes, sorry, i wasn't clear, i want this archive code "import TableCellNodeView from '@/path/to/component';",

**Comment by zanzlender** [comment]
This is on the examples in their docs but I don't know why it's not included...

![image](https://user-images.githubusercontent.com/44570474/220593929-c1aa1be3-9e7e-42ab-8bf8-34acf9206b86.png)

![image](https://user-images.githubusercontent.com/44570474/220593959-f9a33630-850b-419a-a940-086e34c4e609.png)

**Comment by JersyFernandesJF** [comment]
@jondavidpass, i have the same problem and i dont know how to fix ,  can you please share the code for it ?

**Reply by jondavidpass** [reply]
The code I'm using to select cells is all listed above.

**Reply by jondavidpass** [reply]
@JersyFernandesJF 

There's lots of pretty specific UI stuff in there.

The docs for node views are here: https://tiptap.dev/guide/node-views
And the important code to include is as above:
```
    methods: {
        selectRow() {
            this.editor.chain().focus().selectRow(this.getPos()).run();
        },
        selectColumn() {
            this.editor.chain().focus().selectColumn(this.getPos()).run();
        }
    }
```
